lib/elixir/lib/exception.ex

```diff
@@ -157,6 +157,159 @@ defmodule Exception do
+    end
+  end
+
+  def blame(_kind, reason, stacktrace) do
+    {reason, stacktrace}
+  end
+
+  defp blame_function_clause_error(%{module: module, function: function, arity: arity} = exception,
+                                   [{module, function, args, meta} | rest])
+       when length(args) == arity do
+    exception =
+      case blame_mfa(module, function, args) do
+        {:ok, kind, clauses} -> %{exception | args: args, kind: kind, clauses: clauses}
+        :error -> %{exception | args: args}
+      end
+    {exception, [{module, function, arity, meta} | rest]}
+  end
+  defp blame_function_clause_error(exception, stacktrace) do
+    {exception, stacktrace}
+  end
+
+  @doc """
+  Blames the invocation of the given module, function and arguments.
+
+  This function will retrieve the available clauses from bytecode
+  and evaluate them against the given arguments. The clauses are
+  returned as a list of `{args, guards}` pairs where each argument
+  and each top-level condition in a guard separated by `and`/`or`
+  is wrapped in a tuple with blame metadata.
+
+  This function returns either `{:ok, kind, clauses}` or `:error`.
+  Where `kind` is `:def` or `:defp`. Note this functionality requires
+  Erlang/OTP 20, otherwise `:error` is always returned.
+  """
+  @spec blame_mfa(module, function, args :: [term]) ::
+        {:ok, :def | :defp, [{args :: [term], guards :: [term]}]} | :error
+  def blame_mfa(module, function, args) when is_atom(module) and is_atom(function) and is_list(args) do
+    try do
+      blame_mfa(module, function, length(args), args)
+    rescue
+      _ -> :error
+    end
+  end
+
+  defp blame_mfa(module, function, arity, call_args) do
+    with path when is_list(path) <- :code.which(module),
+         {:ok, {_, [debug_info: {:debug_info_v1, backend, data}]}} <- :beam_lib.chunks(path, [:debug_info]),
+         {:ok, %{definitions: defs}} <- backend.debug_info(:elixir_v1, module, data, []),
+         {_, kind, _, clauses} <- List.keyfind(defs, {function, arity}, 0) do
+      clauses =
+        for {meta, ex_args, guards, _block} <- clauses do
+          scope = :elixir_erl.definition_scope(meta, kind, function, arity, "nofile")
+          {erl_args, scope} = :elixir_erl_clauses.match(&:elixir_erl_pass.translate_args/2, ex_args, scope)
+          {args, binding} = blame_args(call_args, ex_args, erl_args, [], [])
+          guards = Enum.map(guards, &blame_guard(&1, scope, binding))
+          {args, guards}
+        end
+      {:ok, kind, clauses}
+    else
+      _ -> :error
+    end
+  end
+
+  defp blame_args([call_arg | call_args], [ex_arg | ex_args], [erl_arg | erl_args], binding, ann_args) do
+    {match?, binding} = blame_arg(erl_arg, call_arg, binding)
+    blame_args(call_args, ex_args, erl_args, binding,
+               [blame_wrap(match?, rewrite_arg(ex_arg)) | ann_args])
+  end
+  defp blame_args([], [], [], binding, ann_args) do
+    {Enum.reverse(ann_args), binding}
+  end
```

Comment:
Just throwing it out there and didn't test at all (GitHub-comment-driven-development™), but would replacing the `{args, binding} = blame_args(call_args, ex_args, erl_args, [], [])` line above with the recursive function with a `zip` + `flat_map`? 

```elixir
[call_args, ex_args, erl_args]
|> Enum.zip()
|> Enum.map_reduce([], fn {call_arg, ex_arg, erl_arg}, binding -> some_function_name(call_arg, ex_arg, erl_arg, binding) end)

defp some_function_name(call_arg, ex_arg, erl_args, binding) do
  {match?, binding} = blame_arg(erl_arg, ex_arg, erl_args)
  {blame_wrap(match?, rewrite_arg(ex_arg)), binding}
end
```

?